src/Lean/Server/Requests.lean

```diff
@@ -171,80 +237,70 @@ def withWaitFindSnapAtPos
     (x := f)
 
 open Language.Lean in
-/-- Finds the first `CommandParsedSnapshot` fulfilling `p`, asynchronously. -/
-partial def findCmdParsedSnap (doc : EditableDocument) (p : CommandParsedSnapshot → Bool) :
-    Task (Option CommandParsedSnapshot) := Id.run do
+/-- Finds the first `CommandParsedSnapshot` containing `hoverPos`, asynchronously. -/
+partial def findCmdParsedSnap (doc : EditableDocument) (hoverPos : String.Pos)
+    (config : Language.SnapSelectionConfig) : Task (Option CommandParsedSnapshot) := Id.run do
   let some headerParsed := doc.initSnap.result?
     | .pure none
   headerParsed.processedSnap.task.bind (sync := true) fun headerProcessed => Id.run do
     let some headerSuccess := headerProcessed.result?
       | return .pure none
     headerSuccess.firstCmdSnap.task.bind (sync := true) go
 where
-  go cmdParsed :=
-    if p cmdParsed then
-      .pure (some cmdParsed)
-    else
-      match cmdParsed.nextCmdSnap? with
-      | some next => next.task.bind (sync := true) go
-      | none => .pure none
+  go (cmdParsed : CommandParsedSnapshot) : Task (Option CommandParsedSnapshot) := Id.run do
+    if containsHoverPos cmdParsed then
+      return .pure (some cmdParsed)
+    match cmdParsed.nextCmdSnap? with
+    | some next => next.task.bind (sync := true) go
+    | none => .pure none
```

Comment:
We can return early if `hoverPos` is before `cmdParsed`

Reply:
Good catch!